### PR TITLE
feat(identify): implement signedPeerRecord

### LINKED ISSUE #4017: identify: implement SignedPeerRecord
## Description

As part of the identify protocol `signedPeerRecord` contains a serialized SignedEnvelope containing a PeerRecord,
signed by the sending node. It contains the same addresses as the listenAddrs field, but in a form that lets us share authenticated addrs with other peers.

Currently the `rust-libp2p` identify protocol does not implement `signedPeerRecord`
https://github.com/libp2p/rust-libp2p/blob/master/protocols/identify/src/generated/structs.proto#L5

For reference `go-libp2p` does:
https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/identify/pb/identify.proto#L35

Is there a plan to implement this in the near future, is there something blocking it?

## Motivation

There was an change in `js-libp2p` `v0.45.0` which makes it no longer work the same for identify records where `signedPeerRecord` is omitted https://github.com/libp2p/js-libp2p/issues/1789. This results in there being a requirement for signedPeerRecord in order for peer identify records to be consumed in v0.45.0.

This appears to be a bug on the `js-libp2p` side, however leaving this issue open to track progress of `signedPeerRecord` in the identify protocol. 

## Are you planning to do it yourself in a pull request?

I have implemented something locally and can verify it can work as expected with `js-libp2p` `v0.45.0`. However, I'm not sure I have enough context to contribute to this at this stage.

### PR DESCRIPTION
**Pull request by drHuangMHT**
## Description
May close #4017.

## Notes & open questions   

`pb-rs` now uses `Cow<'_,T>` for the compiled Rust structs. But with borrowed type in the struct, `FramedRead` can no longer process frames correctly(trait bound not statisfied).

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by drHuangMHT** [comment]
Oh CI will compile the protobuf code, didn't see that.  
EDIT: Wait how does this work?

File: protocols/identify/src/behaviour.rs [hunk 1]
```diff
@@ -118,7 +120,7 @@ pub struct Config {
     /// e.g. `ipfs/1.0.0` or `polkadot/1.0.0`.
     protocol_version: String,
     /// The public key of the local node. To report on the wire.
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]: Should we update the comment to reflect on `CryptoKey`?

File: protocols/identify/src/behaviour.rs [hunk 2]
```diff
@@ -156,12 +158,25 @@ pub struct Config {
...
-            local_public_key,
+            local_key: public_key.into(),
+            interval: Duration::from_secs(5 * 60),
+            push_listen_addr_updates: false,
+            cache_size: 100,
+            hide_listen_addrs: false,
+        }
+    }
+
+    pub fn new_with_keypair(protocol_version: String, local_keypair: Keypair) -> Self {
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]: We should probably borrow `Keypair` and clone it within the function, or make a conversion to convert `&Keypair` into `CryptoKey` with it cloning within the From impl. Thoughts on that?

File: protocols/identify/src/behaviour.rs [hunk 3]
```diff
@@ -670,6 +685,39 @@ impl PeerCache {
     }
 }
 
+#[derive(Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub enum CryptoKey {
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]: Since I dont believe this would be used outside of the crate, unlesss `CryptoKey` is expected to be used in some manner outside of the crate too.

```suggestion
pub(crate) enum CryptoKey {
```

**Review by dariusc93 — COMMENTED** [review]
Left some comments :)

File: protocols/identify/src/behaviour.rs (hunk 3 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: That would trigger a visibility warning because `Handler`'s public constructor uses this type, though I don't expect the constructor can actually be called outside of the crate.

File: protocols/identify/src/behaviour.rs (hunk 1 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: Yes the documentation hasn't been updated, will do.

Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: Also the name `CryptoKey` is a weird name but I don't have a better idea, any thoughts?

File: protocols/identify/src/behaviour.rs (hunk 3 above)
Comment (dariusc93) [reply protocols/identify/src/behaviour.rs]: Hmm then we could also mark that as `pub(crate)`?

File: protocols/identify/src/behaviour.rs (hunk 1 above)
Comment (dariusc93) [reply protocols/identify/src/behaviour.rs]: Maybe `KeyType` since we are using either `Keypair` or `PublicKey`?

**Comment by getong** [comment]
you might rebase to main branch, as zlib license has added to ci， #5769

**Comment by getong** [comment]
```
Files in /home/runner/work/rust-libp2p/rust-libp2p/protocols/identify have changed, please write a changelog entry in /home/runner/work/rust-libp2p/rust-libp2p/protocols/identify/CHANGELOG.md
```
I think you might write some changelog text into this file, just like,
```
implement signedPeerRecord
```
the pr might be okay now.

**Comment by drHuangMHT** [comment]
> the pr might be okay now.

Thanks for the review! Working on tests right now.

**Comment by drHuangMHT** [comment]
version `0.45.1` of `libp2p-identify` hasn't been released on crates.io yet, I'm not sure whether I should add a changelog entry right now.

**Comment by elenaf9** [comment]
> version `0.45.1` of `libp2p-identify` hasn't been released on crates.io yet, I'm not sure whether I should add a changelog entry right now.

CHANGELOG will be fixed with #5803, you can just add the new entry to `0.46.0`.

File: protocols/identify/src/behaviour.rs [hunk 4]
```diff
@@ -156,12 +160,29 @@ pub struct Config {
...
+            cache_size: 100,
+            hide_listen_addrs: false,
+        }
+    }
+
+    /// Creates a new configuration for the identify [`Behaviour`] that
+    /// advertises the given protocol version and public key.  
+    /// The private key will be used to sign [`PeerRecord`](libp2p_core::PeerRecord)
+    /// for verifiable address advertisement.
+    pub fn new_with_keypair(protocol_version: String, local_keypair: &Keypair) -> Self {
```
Comment (elenaf9) [inline protocols/identify/src/behaviour.rs]:
```suggestion
    pub fn new_with_signed_peer_records(protocol_version: String, local_keypair: &Keypair) -> Self {
```

File: protocols/identify/src/behaviour.rs [hunk 5]
```diff
@@ -156,12 +160,29 @@ pub struct Config {
...
-    /// advertises the given protocol version and public key.
-    pub fn new(protocol_version: String, local_public_key: PublicKey) -> Self {
+    /// advertises the given protocol version and public key.  
+    /// Use [`new_with_keypair`](Config::new_with_keypair) for `signedPeerRecord` support.  
+    pub fn new(protocol_version: String, public_key: PublicKey) -> Self {
         Self {
             protocol_version,
             agent_version: format!("rust-libp2p/{}", env!("CARGO_PKG_VERSION")),
-            local_public_key,
+            local_key: public_key.into(),
```
Comment (elenaf9) [inline protocols/identify/src/behaviour.rs]:
```suggestion
    pub fn new(protocol_version: String, local_public_key: PublicKey) -> Self {
        Self {
            protocol_version,
            agent_version: format!("rust-libp2p/{}", env!("CARGO_PKG_VERSION")),
            local_key: local_public_key.into(),
```
Stick with `local_public_key` so it's consistent with `local_keypair` below?

**Review by elenaf9 — CHANGES_REQUESTED** [review]
Thank you @drHuangMHT! Looks like a great start.

I found a pending libp2p spec for signed peer records: https://github.com/libp2p/specs/pull/630. Most of it matches this PR, just one thing is missing: 

> If the `signedPeerRecord` is present the implementation MUST use the data contained within it and ignore duplicated fields present in the main identify message

IMO that makes sense and should be added to this PR. So instead of copying the received record as it is in `handle_incoming_info`, I think we should instead (when an signed envelope is present):
1. Verify that the signature matches the remote's public key
2. Deserialize the record, store the resulting addresses as `listen_addrs`, and ignore the original `listen_addrs`.

Wdyt?

File: protocols/identify/src/behaviour.rs (hunk 5 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: I would like to keep existing API unchanged, so I'm changing `local_keypair` to `keypair` below.

Comment (elenaf9) [reply protocols/identify/src/behaviour.rs]: But isn't it called `local_public_key` in the existing API?

File: protocols/identify/src/protocol.rs [hunk 6]
```diff
@@ -226,16 +226,29 @@ impl TryFrom<proto::Identify> for Info {
...
+        // When signedPeerRecord contains valid addresses, ignore addresses in listenAddrs.  
+        // When signedPeerRecord is invalid or signed by others, ignore the signedPeerRecord(set to `None`).  
+        let (signed_peer_record, listen_addrs) = signed_peer_record
+            .as_ref()
+            .and_then(|envelope| PeerRecord::try_deserialize_signed_envelope(&envelope).ok())
+            .and_then(|(envelope_public_key, _, _, addresses)| {
+                (*envelope_public_key == public_key).then_some(addresses)
+            })
+            .map(|addrs| (signed_peer_record, addrs))
+            .unwrap_or_else(|| (None, parse_listen_addrs(msg.listenAddrs)));
```
Comment (elenaf9) [inline protocols/identify/src/protocol.rs]: Do we really need to `deserialize` the record again? Can't we just read out `PeerRecord::adddresses`:
```rs
let listen_addrs = signed_peer_record.map_or_else(
     || parse_listen_addrs(msg.listenAddrs), 
     |record| record.addresses.to_vec()
);

```

Comment (elenaf9) [reply protocols/identify/src/protocol.rs]: Ahh we also need to check that the envelope key matches, but that can be done as well by reading the `PeerRecord::peer_id` and comparing it with the remote's id, right?

Comment (drHuangMHT) [reply protocols/identify/src/protocol.rs]: > Do we really need to `deserialize` the record again? Can't we just read out `PeerRecord::adddresses`:
> 
> ```rust
> let listen_addrs = signed_peer_record.map_or_else(
>      || parse_listen_addrs(msg.listenAddrs), 
>      |record| record.addresses.to_vec()
> );
> ```

`record` has type `SignedEnvelope`, whose payload is in the form of `Vec<u8>`(bytes) that doesn't have `addresses` field unless we deserialize the payload into `PeerRecord`. In order to own the addresses the only way is to call `record.addresses().to_vec()` which in this case allocates twice, which is unnecessary. So I extracted `PeerRecord::try_deserialize_signed_envelope` to cut down allocation.

Comment (drHuangMHT) [reply protocols/identify/src/protocol.rs]: > Ahh we also need to check that the envelope key matches, but that can be done as well by reading the `PeerRecord::peer_id` and comparing it with the remote's id, right?

`PeerId` is derived from a public key, which *indirectly* proves its identity while the key itself does so directly. ID can collide, but the key is less likely, considering we support multiple key types. I believe it is safer to compare the keys directly when they are already present.

Comment (elenaf9) [reply protocols/identify/src/protocol.rs]: _Sorry for the late reply!_

---

> `record` has type `SignedEnvelope`, whose payload is in the form of `Vec<u8>`(bytes) that doesn't have `addresses` field unless we deserialize the payload into `PeerRecord`.

Ah I see. The name `signed_peer_record` had me confused. How about calling it `signed_envelope` instead, so it matches the type (also in `Info`)?

> In order to own the addresses the only way is to call `record.addresses().to_vec()` which in this case allocates twice, which is unnecessary. So I extracted `PeerRecord::try_deserialize_signed_envelope` to cut down allocation.

Good point. Still, I would expect that the compiler optimizes away the unnecessary allocation. IMO the current code is a bit difficult to understand, and for the sake of simplicity it would be worth it to do the conversion to `PeerRecord` so that we can then simply clone `PeerRecord::addresses`. 
However, if you have a strong preference for the current implementation I am okay with it. But then I think `PeerRecord::try_deserialize_signed_envelope` logic better fits as `SignedEnvelope::try_deserialize(&self) -> ...` (same logic, just move to `SignedEnvelope`).

> `PeerId` is derived from a public key, which indirectly proves its identity while the key itself does so directly. ID can collide, but the key is less likely, considering we support multiple key types. I believe it is safer to compare the keys directly when they are already present.

We rely on the collision resistance of `PeerId` throughout all of rust-libp2p, and [go-libp2p also only compares IDs here](https://github.com/libp2p/go-libp2p/blob/4957d357e00a15fa24dba4ab21d890f4983473f6/p2p/protocol/identify/id.go#L877-L879). I'm not against comparing public keys, but given that `PeerRecord::try_deserialize_signed_envelope` already compares the `PeerId`s I don't think we need the double check here.

Comment (drHuangMHT) [reply protocols/identify/src/protocol.rs]: > However, if you have a strong preference for the current implementation I am okay with it. But then I think `PeerRecord::try_deserialize_signed_envelope` logic better fits as `SignedEnvelope::try_deserialize(&self) -> ...` (same logic, just move to `SignedEnvelope`).

I don't think it is a good idea to implement deserialization on `SignedEnvelope` because it can contain arbitrary payload, including but not limited to `PeerRecord`. We need specific "headers" to properly deserialize the envelope, according to the spec.

> We rely on the collision resistance of `PeerId` throughout all of rust-libp2p, and [go-libp2p also only compares IDs here](https://github.com/libp2p/go-libp2p/blob/4957d357e00a15fa24dba4ab21d890f4983473f6/p2p/protocol/identify/id.go#L877-L879).

That's fair. Good to know.

> I'm not against comparing public keys, but given that `PeerRecord::try_deserialize_signed_envelope` already compares the `PeerId`s I don't think we need the double check here.

The two checks are different. In `try_deserialize_signed_envelope` it checks whether the signer of the record matches the signer of the envelope. While in `Info::try_from` we essentially check the signer of the envelope(therefore signer of the address record) against the sender of the identify message. If we skip the check in `try_from`, the address can come from peers other than the message sender.

Comment (elenaf9) [reply protocols/identify/src/protocol.rs]: Okay!

File: core/src/peer_record.rs [hunk 7]
```diff
@@ -126,6 +107,30 @@ impl PeerRecord {
     pub fn addresses(&self) -> &[Multiaddr] {
         self.addresses.as_slice()
     }
+
+    pub fn try_deserialize_signed_envelope(
```
Comment (elenaf9) [inline core/src/peer_record.rs]: Please document the return parameters.

File: protocols/identify/src/protocol.rs [hunk 8]
```diff
@@ -213,21 +218,37 @@ impl TryFrom<proto::Identify> for Info {
...
+        // When signedPeerRecord contains valid addresses, ignore addresses in listenAddrs.
+        // When signedPeerRecord is invalid or signed by others, ignore the signedPeerRecord(set to `None`).
+        let (signed_envelope, listen_addrs) = signed_envelope
+            .as_ref()
+            .and_then(|envelope| PeerRecord::try_deserialize_signed_envelope(envelope).ok())
+            .and_then(|(envelope_public_key, _, _, addresses)| {
+                (*envelope_public_key == identify_public_key).then_some(addresses)
+            })
+            .map(|addrs| (signed_envelope, addrs))
+            .unwrap_or_else(|| (None, parse_listen_addrs(msg.listenAddrs)));
```
Comment (elenaf9) [inline protocols/identify/src/protocol.rs]:
```suggestion
        // When signedPeerRecord contains valid addresses, ignore addresses in listenAddrs.
        // When signedPeerRecord is invalid or signed by others, ignore the signedPeerRecord(set to `None`).
        let (signed_envelope, listen_addrs) = msg
            .signedPeerRecord
            .and_then(|b| {
                let envelope = SignedEnvelope::from_protobuf_encoding(b.as_ref()).ok()?;
                let (envelope_public_key, _, _, addresses) =
                    PeerRecord::try_deserialize_signed_envelope(&envelope).ok()?;
                (*envelope_public_key == identify_public_key).then_some((Some(envelope), addresses))
            })
            .unwrap_or_else(|| (None, parse_listen_addrs(msg.listenAddrs)));
```

File: protocols/identify/src/behaviour.rs [hunk 9]
```diff
@@ -670,6 +691,39 @@ impl PeerCache {
...
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum KeyType {
+    // With public key only the behaviour will not
+    // be able to produce a `SignedEnvelope`.
+    PublicKey(PublicKey),
+    Keypair {
+        keypair: Keypair,
+        public_key: PublicKey,
+    },
+}
```
Comment (elenaf9) [inline protocols/identify/src/behaviour.rs]: Would it make sense to heap allocate the `Keypair` here (i.e. `Box<Keypair>`) to avoid the large-enum-variant issue?

Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: I don't have an opinion on this. ~~Might not be necessary since only the `Behaviour` holds one instance of this enum.~~

File: protocols/identify/src/behaviour.rs [hunk 10]
```diff
@@ -670,6 +691,39 @@ impl PeerCache {
...
+    // Reduce enum size with heap allocated `Box<T>`
+    PublicKey(Box<PublicKey>),
+    Keypair {
+        keypair: Box<Keypair>,
+        public_key: Box<PublicKey>,
+    },
+}
+impl From<PublicKey> for KeyType {
+    fn from(value: PublicKey) -> Self {
+        Self::PublicKey(value.clone().into())
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]:
```suggestion
        Self::PublicKey(value.into())
```

File: protocols/identify/src/behaviour.rs [hunk 11]
```diff
@@ -156,12 +160,29 @@ pub struct Config {
 
 impl Config {
     /// Creates a new configuration for the identify [`Behaviour`] that
-    /// advertises the given protocol version and public key.
+    /// advertises the given protocol version and public key.  
+    /// Use [`new_with_keypair`](Config::new_with_keypair) for `signedPeerRecord` support.
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]:
```suggestion
    /// Use [`new_with_signed_peer_record`](Config:: new_with_signed_peer_record) for `signedPeerRecord` support.  
```

File: protocols/identify/src/behaviour.rs (hunk 9 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: Since we are duplicating this in every `ConnectionHandler`, I think we can reference counting it. What do you think @elenaf9?

Comment (elenaf9) [reply protocols/identify/src/behaviour.rs]: Yes sounds good :+1:

Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: In that case do we still box the data inside? If so, would it be better if we use a tuple or create a struct for semantic purpose?

Comment (elenaf9) [reply protocols/identify/src/behaviour.rs]: > In that case do we still box the data inside?

No need to `Box` it, `Arc` already allocates the data on the heap.

Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: > No need to `Box` it, `Arc` already allocates the data on the heap.

I know but the `Box` was supposed to get rid of `large_enum_variant` right? Does it make sense to have two separate `Arc`?

Comment (elenaf9) [reply protocols/identify/src/behaviour.rs]: > I know but the `Box` was supposed to get rid of `large_enum_variant` right? Does it make sense to have two separate `Arc`?

I am not sure I know what you mean.
My idea was to just make the `keypair: Keypair` propterty a `Box<Keypair>`/ `Arc<Keypair>` and leave `public_key: PublicKey` as it is. That way the keypair would be on the heap and the enum only stores the pointer, thus avoiding the `large_enum_variant`.

Alternatively we could also not have this enum, stick with the old `local_public_key: PublicKey` and add have an additional `records_signing_key: Option<Keypair>`.

Comment (dariusc93) [reply protocols/identify/src/behaviour.rs]: I think it might be best to go for the alt Elena mentioned if we are to avoid using Box/Arc. Additionally if we are to use it, we could use a struct that as an inner enum that is wrapped with `Arc` but at that point it would probably be better to just have a struct with two fields with one being a public key and another an optional keypair (though at that point we are getting back to what was mentioned above).

File: protocols/identify/src/behaviour.rs [hunk 12]
```diff
@@ -156,12 +160,30 @@ pub struct Config {
...
         Self {
             protocol_version,
             agent_version: format!("rust-libp2p/{}", env!("CARGO_PKG_VERSION")),
-            local_public_key,
+            local_key: local_public_key.into(),
+            interval: Duration::from_secs(5 * 60),
+            push_listen_addr_updates: false,
+            cache_size: 100,
+            hide_listen_addrs: false,
+        }
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]: I wonder if we could move the body from `new` and `new_with_signed_peer_record` into its own private function so we could reduce duplicated code. Thoughts?

File: protocols/identify/src/behaviour.rs (hunk 9 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: Let me rephrase. First, we would like to get rid of `large_enum_variant` lint using heap allocation. Then I tried to box all fields(misread your message) but I didn't think allocate twice is very optimal so I asked should we do `Box<BothKey>` or `Box<(Keypair, PublicKey)>`. Plus we are duplicating `KeyType` the enum in every `ConnectionHandler`, so I asked "should we reference counting it(`KeyType`)" so that there would only be one copy of `KeyType` and the lint won't really matter. I guess what you mean is reference counting one field but still cloning `KeyType`. I think it would be better if we just reference count the entire enum.

File: protocols/identify/src/behaviour.rs [hunk 13]
```diff
@@ -156,12 +161,26 @@ pub struct Config {
...
+
+    /// Creates a new configuration for the identify [`Behaviour`] that
+    /// advertises the given protocol version and public key.  
+    /// The private key will be used to sign [`PeerRecord`](libp2p_core::PeerRecord)
+    /// for verifiable address advertisement.
+    pub fn new_with_signed_peer_record(protocol_version: String, local_keypair: &Keypair) -> Self {
+        Self::new_with_key(protocol_version, local_keypair.into())
+    }
+
+    fn new_with_key(protocol_version: String, key_type: KeyType) -> Self {
```
Comment (dariusc93) [inline protocols/identify/src/behaviour.rs]: nit: We could use `impl Into<KeyType>` and then convert it there in the function so that way we can just pass `Keypair` or `PublicKey` into the function.

File: protocols/identify/src/behaviour.rs (hunk 9 above)
Comment (elenaf9) [reply protocols/identify/src/behaviour.rs]: > I think it would be better if we just reference count the entire enum

Fine by me as well. I think the increase enum size is okay if it's only allocated once.

File: protocols/identify/src/behaviour.rs [hunk 14]
```diff
@@ -117,8 +120,10 @@ pub struct Config {
     /// Application-specific version of the protocol family used by the peer,
     /// e.g. `ipfs/1.0.0` or `polkadot/1.0.0`.
     protocol_version: String,
-    /// The public key of the local node. To report on the wire.
-    local_public_key: PublicKey,
+    /// The key of the local node. Only the public key will be report on the wire.  
+    /// The behaviour will not produce [`PeerRecord`](libp2p_core::PeerRecord) when
+    /// supplied with a public key.  
+    local_key: Arc<KeyType>,
```
Comment (elenaf9) [inline protocols/identify/src/behaviour.rs]:
```suggestion
    /// The behaviour will only send signed [`PeerRecord`](libp2p_core::PeerRecord)s in
    /// its identify message for `KeyType::Keypair`.
    local_key: Arc<KeyType>,
```

**Review by elenaf9 — APPROVED** [review]
LGTM. Thanks @drHuangMHT for the follow-ups!

I'd still like to wait for a review from @jxs before merging this.

File: protocols/identify/src/behaviour.rs (hunk 13 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: No. That will trigger a visibility warning if we don't make `KeyType` public.

File: protocols/identify/src/behaviour.rs (hunk 10 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: Changes reverted because we're now reference-counting the entire enum.

File: protocols/identify/src/behaviour.rs (hunk 13 above)
Comment (dariusc93) [reply protocols/identify/src/behaviour.rs]: That shouldn't be a problem since we would only be using `impl Into<KeyType>` on `new_with_key` which is a private function, unless im missing something?

Comment (dariusc93) [reply protocols/identify/src/behaviour.rs]: I.e this should not give a warning

```diff
diff --git a/protocols/identify/src/behaviour.rs b/protocols/identify/src/behaviour.rs
index da0d4a510..e22afabd1 100644
--- a/protocols/identify/src/behaviour.rs
+++ b/protocols/identify/src/behaviour.rs
@@ -165,7 +165,7 @@ impl Config {
     /// Use [`new_with_signed_peer_record`](Config::new_with_signed_peer_record) for
     /// `signedPeerRecord` support.
     pub fn new(protocol_version: String, local_public_key: PublicKey) -> Self {
-        Self::new_with_key(protocol_version, local_public_key.into())
+        Self::new_with_key(protocol_version, local_public_key)
     }
 
     /// Creates a new configuration for the identify [`Behaviour`] that
@@ -173,10 +173,11 @@ impl Config {
     /// The private key will be used to sign [`PeerRecord`](libp2p_core::PeerRecord)
     /// for verifiable address advertisement.
     pub fn new_with_signed_peer_record(protocol_version: String, local_keypair: &Keypair) -> Self {
-        Self::new_with_key(protocol_version, local_keypair.into())
+        Self::new_with_key(protocol_version, local_keypair)
     }
 
-    fn new_with_key(protocol_version: String, key_type: KeyType) -> Self {
+    fn new_with_key(protocol_version: String, key_type: impl Into<KeyType>) -> Self {
+        let key_type = key_type.into();
         Self {
             protocol_version,
             agent_version: format!("rust-libp2p/{}", env!("CARGO_PKG_VERSION")),
```

but this isnt required though and shouldnt block the pr from being merged :)

File: protocols/identify/src/protocol.rs [hunk 15]
```diff
@@ -53,6 +53,8 @@ pub struct Info {
     pub protocols: Vec<StreamProtocol>,
     /// Address observed by or for the remote.
     pub observed_addr: Multiaddr,
+    /// Verifiable addresses of the peer.
+    pub signed_peer_record: Option<SignedEnvelope>,
```
Comment (dariusc93) [inline protocols/identify/src/protocol.rs:57]: since this is adding a new field to the struct, it breaks semver unless we bump the minor version.

File: protocols/identify/src/behaviour.rs (hunk 13 above)
Comment (drHuangMHT) [reply protocols/identify/src/behaviour.rs]: > I.e this should not give a warning

I put the opaque type on public constructor and that triggered a visibility warning, but yeah it is fine to put it on the private constructor.

File: core/src/peer_record.rs [hunk 16]
```diff
@@ -30,26 +30,7 @@ impl PeerRecord {
...
-        }
-
-        let seq = record.seq;
-        let addresses = record
-            .addresses
-            .into_iter()
-            .map(|a| a.multiaddr.to_vec().try_into())
-            .collect::<Result<Vec<_>, _>>()?;
-
+        let (_, peer_id, seq, addresses) = Self::try_deserialize_signed_envelope(&envelope)?;
```
Comment (jxs) [inline core/src/peer_record.rs]: `try_deserialize_signed_envelope` doesn't feel like a simplification to me, we are `pub`ing a function that returns a tuple of four members and incrementing `lip2p-core` with all its inherent issues, whereas `from_signed_envelope` returns the `PeerRecord` with the same fields, can't we just use it?

File: protocols/identify/src/generated/structs.proto [hunk 17]
```diff
@@ -24,4 +24,11 @@ message Identify {
   optional bytes observedAddr = 4;
 
   repeated string protocols = 3;
+
+  // signedPeerRecord contains a serialized SignedEnvelope containing a PeerRecord,
+  // signed by the sending node. It contains the same addresses as the listenAddrs field, but
+  // in a form that lets us share authenticated addrs with other peers.
+  // see github.com/libp2p/go-libp2p/core/record/pb/envelope.proto and
+  // github.com/libp2p/go-libp2p/core/peer/pb/peer_record.proto for message definitions.
```
Comment (jxs) [inline protocols/identify/src/generated/structs.proto]: these links both return 404

File: protocols/identify/src/protocol.rs [hunk 18]
```diff
@@ -293,10 +317,87 @@ mod tests {
...
+            0x62, 0x70, 0x32, 0x70, 0x2f, 0x67, 0x6f, 0x2d, 0x6c, 0x69, 0x62, 0x70, 0x32, 0x70,
+            0x2f, 0x70, 0x32, 0x70, 0x2f, 0x70, 0x72, 0x6f, 0x74, 0x6f, 0x63, 0x6f, 0x6c, 0x2f,
+            0x69, 0x64, 0x65, 0x6e, 0x74, 0x69, 0x66, 0x79, 0x2f, 0x70, 0x62,
+        ];
+        let mut buf = [0u8; 375];
+        let mut message =
+            proto::Identify::from_reader(&mut BytesReader::from_bytes(&go_protobuf), &go_protobuf)
+                .expect("read to succeed");
+
+        // The actual bytes they put in is "github.com/libp2p/go-libp2p/p2p/protocol/identify/pb".
```
Comment (jxs) [inline protocols/identify/src/protocol.rs:370]: this also returns a 404, I know this code is copied, but we shouldn't be adding broken addresses if we know they are broken

**Review by jxs — COMMENTED** [review]
Hi Dr Huang, overall looks good to me, thanks fr your work on this!
let some comments

File: protocols/identify/src/generated/structs.proto (hunk 17 above)
Comment (dariusc93) [reply protocols/identify/src/generated/structs.proto]: imho i think it might also be worth including the commit in the links as well in case things changes in those repo.

File: protocols/identify/src/protocol.rs (hunk 18 above)
Comment (drHuangMHT) [reply protocols/identify/src/protocol.rs:370]: No, this one is actually a plain string.

File: core/src/peer_record.rs (hunk 16 above)
Comment (drHuangMHT) [reply core/src/peer_record.rs]: We will have to clone the addresses. We can do that if it is acceptable, see the discussion [here](https://github.com/libp2p/rust-libp2p/pull/5785#discussion_r1912309888).

File: protocols/identify/src/generated/structs.proto (hunk 17 above)
Comment (drHuangMHT) [reply protocols/identify/src/generated/structs.proto]: I'm linking them to [here](https://github.com/libp2p/rust-libp2p/blob/8ac5b5aac5f5c25a85f1065e292deeaf58290189/core/src/generated/peer_record.proto#L11) and [here](https://github.com/libp2p/rust-libp2p/blob/8ac5b5aac5f5c25a85f1065e292deeaf58290189/core/src/generated/envelope.proto#L12)

File: protocols/identify/src/protocol.rs (hunk 18 above)
Comment (jxs) [reply protocols/identify/src/protocol.rs:370]: ah right, thanks for clarifying Dr Huang

File: core/src/peer_record.rs (hunk 16 above)
Comment (jxs) [reply core/src/peer_record.rs]: Sorry missed that conversation as it was already resolved.
why not have a `into_addresses` function then? We already have into functions for the other fields.

Comment (drHuangMHT) [reply core/src/peer_record.rs]: > Sorry missed that conversation as it was already resolved. why not have a `into_addresses` function then? We already have into functions for the other fields.

But we also need to put the envelope back into the message, `into_addresses(self)->Vec<Multiaddr>` will simply drop the envelope unless we clone the envelope beforehand. Does it make sense to introduce a method called `into_addresses_and_envelope(self)->(Vec<Multiaddr>, SignedEnvelope)`?

Comment (jxs) [reply core/src/peer_record.rs]: ah, thanks for explaining Dr Huang.
No let's just `clone` it then please, for the sake of simplification.

**Review by jxs — APPROVED** [review]
LGTM! Thanks Dr Huang :rocket:
